### PR TITLE
OWLS-93071 - Merge changes to display correct server status and counts when serverStartPolicy switched to NEVER

### PR DESCRIPTION
**Pull request by ankedia**
Merge changes from PR 2578 for OWLS-93071.

Use the DPI cached in the DomainProcessorImpl instead of DPI passed when scheduling the domain status updating in DomainStatusUpdater step.

The change also deletes the server pod from DPI when the pod is not found during a list operation and is fully deleted. The server state is displayed as SHUTTING_DOWN when the pod is being deleted and SHUTDOWN after the pod is deleted.

Integration test run is in progress - https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/6958/console

### CODE REVIEW

File: operator/src/main/java/oracle/kubernetes/operator/JobWatcher.java [hunk 1]
```diff
@@ -243,7 +244,7 @@ private WaitForJobReadyStep(V1Job job, Step next) {
 
     // A job is considered ready once it has either successfully completed, or been marked as failed.
     @Override
-    boolean isReady(V1Job job) {
+    boolean isReady(V1Job job, DomainPresenceInfo info, String resourceName) {
```
[the remark this answers is not in the document]
Comment (ankedia) [reply operator/src/main/java/oracle/kubernetes/operator/JobWatcher.java]: As discussed, I have added a new step `DomainPresenceInfoUpdateStep` for updating the DPI after pod is fully deleted and reverted previous changes to the JobWatcher, WaitForReady and PodWatcher.

File: operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java [hunk 2]
```diff
@@ -116,7 +116,7 @@ public DomainProcessorImpl(DomainProcessorDelegate delegate, SemanticVersion pro
     this.productVersion = productVersion;
   }
 
-  private static DomainPresenceInfo getExistingDomainPresenceInfo(String ns, String domainUid) {
+  static DomainPresenceInfo getExistingDomainPresenceInfo(String ns, String domainUid) {
```
[the remark this answers is not in the document]
Comment (ankedia) [reply operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java]: Russ and I paired up and added new unit test `afterChangeToNever_statusUpdateRetainsDesiredState` for the scheduling code that reproduces this issue. In the revised fix, status scheduler fetches the DPI from DOMAINS when needed using UID and namespace.

File: operator/src/main/java/oracle/kubernetes/operator/DomainStatusUpdater.java [hunk 3]
```diff
@@ -278,19 +281,19 @@ DomainStatus getNewStatus() {
     }
 
     String getDomainUid() {
-      return getDomain().getDomainUid();
+      return Optional.ofNullable(getDomain()).map(Domain::getDomainUid).orElse(info.getDomainUid());
```
[the remark this answers is not in the document]
Comment (ankedia) [reply operator/src/main/java/oracle/kubernetes/operator/DomainStatusUpdater.java]: Fixed.

**Comment by ankedia** [comment]
@russgold @rjeberhard Can you please review this PR and let me know if any other changes are needed? My other merge related changes are dependent on this branch so it'll be good to get this reviewed and merged if possible. Thanks.

**Review by russgold — APPROVED** [review]

**Review by jshum2479 — APPROVED** [review]

**Pull request merged**